next-app/lib/messenger.js

```diff
@@ -0,0 +1,72 @@
+import { EventEmitter } from "./eventEmitter.js";
+
+export class Messenger extends EventEmitter {
+  #queue = [];
+  #origin;
+  #defaultRecipient;
+  #isAllowed;
+
+  constructor(origin, defaultRecipient, isAllowed = true) {
+    super();
+    this.#origin = origin;
+    this.#defaultRecipient = defaultRecipient;
+    this.#isAllowed = isAllowed;
+  }
+
+  static directSend(origin, recipient, message) {
+    const messenger = new Messenger(origin, recipient);
+    messenger.send(message);
+  }
+
+  listen() {
+    chrome.runtime.onMessage.addListener((message, sender) => {
+      if(message.recipient !== this.#origin) return;
+      this.#receive(message);
+    });
+  }
+
+  setIsAllowed(isAllowed) {
+    this.#isAllowed = isAllowed;
+    this.#dispatch();
+  }
+
+  #dispatch() {
+    for (const message of this.#queue) {
+      let recipient;
+      if (message.recipient) {
+        recipient = message.recipient;
+        delete message.recipient;
+      } else {
+        recipient = this.#defaultRecipient;
+      }
+      /* console.log(`%cmessage recipient:`, "color: green;");

```

Comment:
do we need this comment for smth?

Reply:
i had it for debugging (because i changed logic how the message gets sent/delivered) so it is now obsoloete - i will remove it